Add dictGetAll function to collect multiple matches in RegExpTreeDictionary

johanngan opened the pull request:
This function outputs an array of attribute values from all regexp nodes that matched in a regexp tree dictionary. An optional final argument can be passed to limit the array size.

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added the dictGetAll function for regexp tree dictionaries to return values from multiple matches as arrays. Closes https://github.com/ClickHouse/ClickHouse/issues/50254.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

UnamedRus commented:
May be it make sense to have separate dictGet function for that? Like for hierarchical dictionaries?

johanngan commented:
> May be it make sense to have separate dictGet function for that? Like for hierarchical dictionaries?

Ooh that's an interesting suggestion. I haven't used hierarchical dictionaries and didn't notice the similarities. I'm not sure how to implement a new dictGet function, but I'll look into it.

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #50255 from johanngan/ClickHouse:regexptree-arrays -->
This is an automated comment for commit af0869eaf36570ea7d852def6e9413d4c3bf3d89 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50255/af0869eaf36570ea7d852def6e9413d4c3bf3d89/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

johanngan committed:
Add dictGetAll function for RegExpTreeDictionary

This function outputs an array of attribute values from all regexp nodes
that matched in a regexp tree dictionary. An optional final argument can
be passed to limit the array size.

johanngan commented:
Implemented the suggestion. This PR now adds a separate `dictGetAll` function instead of a dictionary setting. Ready for review.

UnamedRus commented:
Minor thing about parameter, JFYI

ClickHouse have special syntax for them (and they also constant):
```
groupArraySample(max_size[, seed])(x)
```
https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/grouparraysample

Not sure what is better having it as attribute or parameter

johanngan commented:
> Minor thing about parameter, JFYI
> 
> ClickHouse have special syntax for them (and they also constant):
> 
> ```
> groupArraySample(max_size[, seed])(x)
> ```
> 
> https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/grouparraysample
> 
> Not sure what is better having it as attribute or parameter

Yeah I actually thought about this too. But AFAICT none of the "regular functions" use this syntax; I only see it used with the aggregate functions. So I think a normal parameter makes more sense here. (I also didn't know how to implement the special parametric syntax)

UnamedRus commented:
> But AFAICT none of the "regular functions" use this syntax

AFAIK, you are right.

> So I think a normal parameter makes more sense here.

Yeah, i'm don't have any strong opinion on that topic

May be @alexey-milovidov can say his opinion on that.

And probably it would be nice to have some article with style guidelines for function arguments/parameters.


dictGetAll vs dictGetMatches vs dictGet? 

Also, All seems a bit "Wide" for me, but again no strong opinion on that and cannot suggest better name :)

rschu1ze commented:
@hanfei1991

hanfei1991 commented on src/Functions/FunctionsExternalDictionaries.h:
```diff
@@ -464,6 +484,19 @@ class FunctionDictGetNoType final : public IFunction
+            if (dictionary_get_function_type == DictionaryGetFunctionType::getAll && current_arguments_index < arguments.size())
+            {
+                auto limit_col = arguments[current_arguments_index].column;
+                if (!limit_col || !isColumnConst(*limit_col))
```
Do we need to check if it is uint as well?

johanngan replied:
I don't think so since the call to `getUInt` right after this already throws an exception if we can't convert to a uint. It's also nice because `getUInt` will cast other arithmetic types to uint, so it still works if I pass other integer types besides uint64.

I added a comment; let me know if that's insufficient.

hanfei1991 commented on src/Functions/FunctionsExternalDictionaries.h:
```diff
@@ -360,6 +370,15 @@ class FunctionDictGetNoType final : public IFunction
         }
 
         bool key_is_nullable = arguments[2].type->isNullable();
+        if (dictionary_get_function_type == DictionaryGetFunctionType::getAll)
```
constexpr?

johanngan replied:
Done

hanfei1991 commented on src/Functions/FunctionsExternalDictionaries.h:
```diff
@@ -565,31 +599,40 @@ class FunctionDictGetNoType final : public IFunction
-                key_types,
-                default_cols);
+            Columns result_columns;
+            if (dictionary_get_function_type == DictionaryGetFunctionType::getAll)
```
constexpr?

johanngan replied:
Done.

johanngan committed:
Review comments: Use constexpr-if in more places

Also add a comment about the apparent lack of type checking on the limit
column.

johanngan committed:
Add dictGetAll to spell-check dictionary
